Associations Project

Zubenna opened the pull request:
## In this project, I did the following
- Model and implemented application data relationship.
- Built and migrate User, Event, and Attendance models with associated tables.
- Added simple sign up and login function.
- Created User, Event, and Attendance controllers with the corresponding routes.
- Added page Navigation.
- Set up RSpec and Capybara test environment. Added necessary tests.

Zubenna committed:
implemented data modelling, added tables

Zubenna committed:
Adding user sign-in form

Zubenna committed:
Added user signup, login, and logout

Zubenna committed:
Added display for all, upcoming and past events

Zubenna committed:
Added navigation link to all events and new event

Zubenna committed:
modified user show page to display past and upcoming events

Zubenna committed:
Added events to attendees table, added indexes

Zubenna committed:
Added flash touser and event  create, destroy actions

Zubenna committed:
Set up test environment with Rspec and capybara

Zubenna committed:
Added feature test for user, event and model test

Zubenna committed:
Updated Readme.MD file, adding test procedures

Zubenna committed:
Removed unecessary assignment, spaces and line from code

Zubenna committed:
Corrected identation error in spec/spec_helper.rb file

Zubenna committed:
Removed extra space from spec/model_spec.rb file

Zubenna committed:
Rearranged identation in spec/spec_helper.rb

Zubenna committed:
Realligned identation in spec/spec_helper.rb

Zubenna committed:
corrected identation error

Zubenna committed:
Arranged some gem alphabetically in their group

Zubenna committed:
Removed comments from config/routes.rb file

voscarmv reviewed (changes requested):
# Nice work so far, team. 👍 
## However, there are some changes required 👀 🧰 🔨 🔧 

I cannot begin to review your project as the migrations are broken. Please roll back all migrations and try to migrate again. You should see the following error when you attempt to `rails rb:migrate`:

```
oscar@orion:~/tse/private-events$ rails db:migrate
== 20200909115024 CreateUsers: migrating ======================================
-- create_table(:users)
   -> 0.0021s
== 20200909115024 CreateUsers: migrated (0.0022s) =============================

== 20200909160515 CreateAttendances: migrating ================================
-- create_table(:attendances)
   -> 0.0208s
== 20200909160515 CreateAttendances: migrated (0.0210s) =======================

== 20200910014556 RemovePasswordFromUsers: migrating ==========================
-- remove_column(:users, :password, :string)
   -> 0.0413s
== 20200910014556 RemovePasswordFromUsers: migrated (0.0414s) =================

== 20200911221320 CreateEvents: migrating =====================================
-- create_table(:events)
   -> 0.0127s
== 20200911221320 CreateEvents: migrated (0.0133s) ============================

== 20200922121434 AddAttendances: migrating ===================================
-- create_table(:attendances)
rails aborted!
StandardError: An error has occurred, this and all later migrations canceled:

SQLite3::SQLException: table "attendances" already exists
/home/oscar/tse/private-events/db/migrate/20200922121434_add_attendances.rb:3:in `change'
/home/oscar/tse/private-events/bin/rails:9:in `<top (required)>'
/home/oscar/tse/private-events/bin/spring:15:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'

Caused by:
ActiveRecord::StatementInvalid: SQLite3::SQLException: table "attendances" already exists
/home/oscar/tse/private-events/db/migrate/20200922121434_add_attendances.rb:3:in `change'
/home/oscar/tse/private-events/bin/rails:9:in `<top (required)>'
/home/oscar/tse/private-events/bin/spring:15:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'

Caused by:
SQLite3::SQLException: table "attendances" already exists
/home/oscar/tse/private-events/db/migrate/20200922121434_add_attendances.rb:3:in `change'
/home/oscar/tse/private-events/bin/rails:9:in `<top (required)>'
/home/oscar/tse/private-events/bin/spring:15:in `<top (required)>'
bin/rails:3:in `load'
bin/rails:3:in `<main>'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

Once you fix this it will be possible for the next TSE to review your code.

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

Zubenna committed:
Corrected  migration error

Zubenna committed:
Added more test procedure in Readme.md file

Zubenna committed:
Added empty line between methods in model/user.rb

Zubenna committed:
Defined local variable user, in spec/ model_spec.rb

peterrobert commented on app/views/events/new.html.erb:
```diff
@@ -0,0 +1,19 @@
+  </div><br/>
+  <div>
+    <%= f.label :event_date %>
+    <%= f.text_field :event_date %>
```
This should not be a text_field there is actually a field for a date in rails

Zubenna replied:
## The event creation method is working perfectly well on my computer if you enter the date in this format: eg 2020-08-12.  However, I have replaced the text_field with date_select form builder helper method.

peterrobert commented on config/routes.rb:
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  resources :users, only: [:new, :create, :show]
+  resources :events
+  get '/login' => 'sessions#new'
+  post '/login' => 'sessions#create'
+  delete '/logout' => 'sessions#destroy'
+  get '/welcome' => 'sessions#welcome'
```
You can make one of your routes the root route

Zubenna replied:
## The get '/welcome' route has been selected as the root route.
## Navigation links for All events, new event, signed-in user, and logout button for signed-in users has been added. There are All events and new user link for the home page.
# Finally
## Please comment out the _header.html.erb partial in 
## views/layout/application.html.erb first before
## running the RSpec test series to avoid ambiguity

peterrobert reviewed (changes requested):
### CHANGES REQUIRED :recycle: :stop_sign: 
### HELLO  :wave:  @Zubenna GREAT PROGRESS SO FAR. HOWEVER, THERE ARE A FEW CHANGES THAT NEED TO BE MADE KINDLY CHECK BELOW :arrow_down: 

- [ ] The application is creating a user :+1:  However, there seems to be a problem in your event Create method. This is because the data is not getting persisted to the database.

![something](https://user-images.githubusercontent.com/32361963/94356809-2f64d400-009b-11eb-951f-a132e9a9e5a0.png)

- [ ] There should be navigation links at the top of the page.
![navigation](https://user-images.githubusercontent.com/32361963/94356848-a13d1d80-009b-11eb-8623-81e4c31d7ca9.png)

###NB once you fix the above suggestions the next TSE will be able to actually review your project

### If you have any questions kindly reach me on slack @peterrobert
### HAPPY CODING :computer:

Zubenna committed:
Replaced text_field with date_select, added navigation links

Zubenna committed:
Removed trailing white space in spec/event_spec.rb

kobiyoyo reviewed (changes requested):
## Hi Team   ![](https://user-images.githubusercontent.com/38801423/62908122-703d4480-bd44-11e9-9115-42025ab7f341.png) 
#### Great Job so far, everything works fine, but there are a few more things required. 
### General 
- [ ] I get this error, whenever I try to create events.
![1](https://user-images.githubusercontent.com/23278161/94372614-168a0c00-00f7-11eb-800b-6a2058f2439d.png)
- [ ] Navigation links are present at the top of the page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [ ] You should write unit test for all components
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

Zubenna commented:
# I have implemented all these changes even before now. I left clear messages in response to peterrobert's review before kobiyoyo.

## KIndly note the following once again
## 1. The event creation is working perfectly well on my computer. Note that you need to create a user first by clicking on New user or Sign Up. Log into the app with your new username and click on the New event link ontop.

![Eventcreation](https://user-images.githubusercontent.com/15705593/94423319-5b0dba00-0180-11eb-8fba-a07bef90b85c.png)
## 2. If you log into the app, on top of the page is a navigation link comprising All events, New event, and the username of the signed-in user as well as the Logout button below them. Also, note that the emphasis of this project is on functionality, not design. 

## 3. I have added more test cases covering the model, user, and event in the spec folder. Please remember to follow the instruction above the page while running the test to avoid error and ambiguity. See detailed instructions for more tests in the Readme.md page.
Best Regards
Nnamdi.

Zubenna committed:
Replaced New event with New User on the home page

stratospherique reviewed (changes requested):
### Great progress @Zubenna :clap: However, some changes need to be applied in order to get approval for this project. Check :arrow_down: 

#### Implementation (required) :warning: 
* [ ] Some tests are failing. Kindly fix it. Plus, regroup your specs by type (e.g models, feature, etc)
* [ ] Navigation links are missing. The user should be able to visit all the defined routes.
* [ ] When a user is logged in, "log in" and "sign up" button shouldn't be visible.
### That's it :relieved: When finished, kindly submit a code re-review request :pray: 
Happy coding :v: 
**_Ahmed Mahfoudh_**

Zubenna commented:
All requested changes have been implemented.

Zubenna committed:
Seperated tests into folders, added navigation link

Zubenna committed:
Updated the readme.md file with latest information

Zubenna committed:
Corrected identation error in spec/models/attendanc_spec.rb

Zubenna committed:
Updated readme file

kubilaycaglayan commented on app/views/events/show.html.erb:
```diff
@@ -0,0 +1,16 @@
+<p><strong>Event date: </strong><%= @event.event_date %></p>
+<p><strong>Description: </strong><%= @event.description %></p>
+
+<% if @event.attendees.empty? %>
+  <p>No attandees</p>
+<% else %>
+  <strong>Attendees:</strong>
+  <ul>
+    <% @event.attendees.each do |attendee| %>
+      <li><%= link_to attendee.full_name, user_path(attendee.id) %></li>
+    <% end %>
+  </ul>
+<% end %>
+
```
- [ ] [OPTIONAL] There should be absolutely NO logic in views.

kubilaycaglayan reviewed (approved):
Hello Team 👋
Status: Approved :green_circle: 

Great job on making the changes! :blush: 
Everything

All your tests are passing!

Please consider making these optional changes before you move on. These are extremely important details that will help you in the future. ⏬

- [ ] [OPTIONAL] Models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
💡 Consider adding more tests to cover associations.

- [ ] [OPTIONAL] There should be absolutely NO logic in views.
    >Views are only for displaying information that became available to it from its corresponding controller not generating it.

💡 Check the inline comments below please.

Good luck on your next project!
With all my respect...
